### PR TITLE
bash: 5.1 -> 5.2, readline: 8.1 -> 8.2

### PR DESCRIPTION
**Pull request by kaldonir**
###### Description of changes

[Bash 5.2](https://tiswww.case.edu/php/chet/bash/bashtop.html#CurrentStatus) is available since September 2022. This PR bumps the package to the newest version. Bash 5.2 doesn't work with readline 8.1 anymore, so I've also bumped readline to 8.2.

Since a bash and readline update implies many rebuilds, this change is targeting `staging` (not sure if that's correct). I've been unable to verify that *everything* works, but I've built `bashInteractive` and `bash`, where everything seems fine.

There are custom patches for both readline and bash which don't apply to the new versions anymore. I've changed them to apply again, but I don't know what they are actually there for, and whether they are still necessary.

I'm not sure if this is a "Notable version update" that warrants being included in the release notes.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
  - Only for `bashInteractive` and `bash`.
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kaldonir** [comment]
The failed eval-lib-tests works for me locally, and I'm not sure I understand what's happening in the failed logs.

Can somebody help me figure out what I need to do here?

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -13514,16 +13512,16 @@ with pkgs;
 
   any-nix-shell = callPackage ../shells/any-nix-shell { };
 
-  bash = lowPrio (callPackage ../shells/bash/5.1.nix {
+  bash = lowPrio (callPackage ../shells/bash/5.2.nix {
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:13515]:
```suggestion
  bash = lowPrio (callPackage ../shells/bash/5.nix {
```
We might as well remove it to make updates easier

File: pkgs/shells/bash/5.2.nix [hunk 2]
```diff
@@ -55,7 +56,7 @@ stdenv.mkDerivation rec {
   patchFlags = [ "-p0" ];
 
   patches = upstreamPatches
-    ++ [ ./pgrp-pipe-5.1.patch ];
+    ++ [ ./pgrp-pipe-5.2.patch ];
```
Comment (SuperSandro2000) [inline pkgs/shells/bash/5.2.nix:59]:
```suggestion
    ++ [ ./pgrp-pipe-5.patch ];
```

**Comment by SuperSandro2000** [comment]
Not sure either.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by trofi** [comment]
Bisect claims commit 50adabdd60d590c951824974356a9ccb9bb73ffc `readline: 8.1 -> 8.2` broke `gdb`:

```
../../gdb/completer.c: In function 'char* gdb_completion_word_break_characters_throw()':
../../gdb/completer.c:2014:10: error: invalid conversion from 'const char*' to 'char*' [-fpermissive]
 2014 |   return rl_completer_word_break_characters;
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |          |
      |          const char*
```

**Comment by trofi** [comment]
`gdb` is likely fixed upstream by https://sourceware.org/git/?p=binutils-gdb.git;a=commitdiff;h=1add37b567a7dee39d99f37b37802034c3fce9c4

**Comment by trofi** [comment]
Proposed `gdb` fix as https://github.com/NixOS/nixpkgs/pull/208327

**Comment by Artturin** [comment]
`python3.9-hypothesis> /nix/store/63dyk9irl7997cl4i9p1gxk9gr734abv-stdenv-linux/setup: line 129: pop_var_context: head of shell_variables not a function context`

can be reproduced by adding `fail` to a postCheck

https://mail.gnu.org/archive/html/bug-bash/2022-10/msg00075.html

> In this case, it doesn't indicate that anything is wrong. The shell has
flushed all the local variable contexts because it's going to exit due
to errexit being enabled. I'll make a change to suppress the message in
this instance.

so i guess it's only cosmetic

**Comment by SuperSandro2000** [comment]
Did the person already post the patch? If so I think we should apply it to avoid confusion.

**Comment by Artturin** [comment]
we're already applying all the patches
https://git.savannah.gnu.org/cgit/bash.git/log
https://github.com/NixOS/nixpkgs/blob/staging/pkgs/shells/bash/bash-5.2-patches.nix

i guess Chet Ramey forgot to create the patch so we should email him

**Comment by alyssais** [comment]
This broke `pkgsStatic.bash`.  Fix is https://github.com/NixOS/nixpkgs/pull/210480.

**Comment by austinbutler** [comment]
I emailed Chet and he says he made the change on 10/12/2022 and it's in the devel branch. I think he's referring to https://git.savannah.gnu.org/cgit/bash.git/commit/?h=devel&id=055a4552c901e43e6a6007f5bc664a33698dbf85?

**Comment by SuperSandro2000** [comment]
Is that already in a release? If not should we pick it?

**Comment by austinbutler** [comment]
Honestly I don't know. That commit doesn't apply cleanly. I don't actually know if it's the right commit anyways.

**Comment by SuperSandro2000** [comment]
> This broke `pkgsStatic.bash`. Fix is #210480.

Lets catch this next time #215976

**Pull request merged**